Detect attempt to use '...' for struct update

If I try to use "functional record update syntax" (aka "struct update syntax") with three dots instead of two, the compiler could emit a clearer error message.

Given the following code ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=39a57212a40e794a2c10cfec60fdd323)):

```rust
struct V3 {
    x: f32,
    y: f32,
    z: f32
}

fn pz(v: V3) {
    let _ = V3 {
        z: 0.0,
        ...v
    };
}
```

The current output is:

```
error: expected identifier, found `...`
  --> src/lib.rs:10:9
   |
8  |     let _ = V3 {
   |             -- while parsing this struct
9  |         z: 0.0,
10 |         ...v
   |         ^^^ expected identifier

error[E0063]: missing fields `x` and `y` in initializer of `V3`
 --> src/lib.rs:8:13
  |
8 |     let _ = V3 {
  |             ^^ missing `x` and `y`
```

Ideally the output should be more inclusive about what token was expected:

```
error: expected identifier or `..`, found `...`
```

and include a suggestion directing the programmer to the correct syntax:

```
help: to fill in the rest of the fields from `v`, write `..v` (note only two dots)
```

**Motivation**

* Programmers who also write in JavaScript might be familiar with the similar [spread syntax in object literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#spread_in_object_literals), which uses `...`
* In English prose, `...` is more common than `..`

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"chenyukang"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_error_messages/locales/en-US/parser.ftl
+++ b/compiler/rustc_error_messages/locales/en-US/parser.ftl
@@ -112,6 +112,9 @@ parser_missing_semicolon_before_array = expected `;`, found `[`
 parser_invalid_block_macro_segment = cannot use a `block` macro fragment here
     .label = the `block` fragment is within this context
 
+parser_expect_dotdot_not_dotdotdot = expected `..`, found `...`
+    .suggestion = use `..` to fill in the rest of the fields
+
 parser_if_expression_missing_then_block = this `if` expression is missing a block after the condition
     .add_then_block = add a block here
     .condition_possibly_unfinished = this binary operation is possibly unfinished

--- a/compiler/rustc_parse/src/errors.rs
+++ b/compiler/rustc_parse/src/errors.rs
@@ -369,6 +369,15 @@ pub(crate) struct MissingSemicolonBeforeArray {
 }
 
 #[derive(Diagnostic)]
+#[diag(parser_expect_dotdot_not_dotdotdot)]
+pub(crate) struct MissingDotDot {
+    #[primary_span]
+    pub token_span: Span,
+    #[suggestion_verbose(applicability = "maybe-incorrect", code = "..")]
+    pub sugg_span: Span,
+}
+
+#[derive(Diagnostic)]
 #[diag(parser_invalid_block_macro_segment)]
 pub(crate) struct InvalidBlockMacroSegment {
     #[primary_span]

--- a/compiler/rustc_parse/src/errors.rs
+++ b/compiler/rustc_parse/src/errors.rs
@@ -373,7 +373,7 @@ pub(crate) struct MissingSemicolonBeforeArray {
 pub(crate) struct MissingDotDot {
     #[primary_span]
     pub token_span: Span,
-    #[suggestion_verbose(applicability = "maybe-incorrect", code = "..")]
+    #[suggestion(applicability = "maybe-incorrect", code = "..", style = "verbose")]
     pub sugg_span: Span,
 }
 

--- a/compiler/rustc_parse/src/parser/expr.rs
+++ b/compiler/rustc_parse/src/parser/expr.rs
@@ -20,9 +20,9 @@ use crate::errors::{
     InvalidNumLiteralSuffix, LabeledLoopInBreak, LeadingPlusNotSupported, LeftArrowOperator,
     LifetimeInBorrowExpression, MacroInvocationWithQualifiedPath, MalformedLoopLabel,
     MatchArmBodyWithoutBraces, MatchArmBodyWithoutBracesSugg, MissingCommaAfterMatchArm,
-    MissingInInForLoop, MissingInInForLoopSub, MissingSemicolonBeforeArray, NoFieldsForFnCall,
-    NotAsNegationOperator, NotAsNegationOperatorSub, OctalFloatLiteralNotSupported,
-    OuterAttributeNotAllowedOnIfElse, ParenthesesWithStructFields,
+    MissingDotDot, MissingInInForLoop, MissingInInForLoopSub, MissingSemicolonBeforeArray,
+    NoFieldsForFnCall, NotAsNegationOperator, NotAsNegationOperatorSub,
+    OctalFloatLiteralNotSupported, OuterAttributeNotAllowedOnIfElse, ParenthesesWithStructFields,
     RequireColonAfterLabeledExpression, ShiftInterpretedAsGeneric, StructLiteralNotAllowedHere,
     StructLiteralNotAllowedHereSugg, TildeAsUnaryOperator, UnexpectedTokenAfterLabel,
     UnexpectedTokenAfterLabelSugg, WrapExpressionInParentheses,
@@ -2897,6 +2897,21 @@ impl<'a> Parser<'a> {
                 }
                 self.recover_struct_comma_after_dotdot(exp_span);
                 break;
+            } else if self.token == token::DotDotDot {
+                // suggest `..v` instead of `...v`
+                let snapshot = self.create_snapshot_for_diagnostic();
+                let span = self.token.span;
+                self.bump();
+                match self.parse_expr() {
+                    Ok(_p) => {
+                        self.sess.emit_err(MissingDotDot { token_span: span, sugg_span: span });
+                        break;
+                    }
+                    Err(inner_err) => {
+                        inner_err.cancel();
+                        self.restore_snapshot(snapshot);
+                    }
+                }
             }
 
             let recovery_field = self.find_struct_error_after_field_looking_code();

--- a/compiler/rustc_parse/src/parser/expr.rs
+++ b/compiler/rustc_parse/src/parser/expr.rs
@@ -2880,7 +2880,7 @@ impl<'a> Parser<'a> {
         };
 
         while self.token != token::CloseDelim(close_delim) {
-            if self.eat(&token::DotDot) {
+            if self.eat(&token::DotDot) || self.recover_struct_fileds_dots(close_delim) {
                 let exp_span = self.prev_token.span;
                 // We permit `.. }` on the left-hand side of a destructuring assignment.
                 if self.check(&token::CloseDelim(close_delim)) {
@@ -2897,21 +2897,6 @@ impl<'a> Parser<'a> {
                 }
                 self.recover_struct_comma_after_dotdot(exp_span);
                 break;
-            } else if self.token == token::DotDotDot {
-                // suggest `..v` instead of `...v`
-                let snapshot = self.create_snapshot_for_diagnostic();
-                let span = self.token.span;
-                self.bump();
-                match self.parse_expr() {
-                    Ok(_p) => {
-                        self.sess.emit_err(MissingDotDot { token_span: span, sugg_span: span });
-                        break;
-                    }
-                    Err(inner_err) => {
-                        inner_err.cancel();
-                        self.restore_snapshot(snapshot);
-                    }
-                }
             }
 
             let recovery_field = self.find_struct_error_after_field_looking_code();
@@ -3040,6 +3025,18 @@ impl<'a> Parser<'a> {
             comma: self.token.span,
         });
         self.recover_stmt();
+    }
+
+    fn recover_struct_fileds_dots(&mut self, close_delim: Delimiter) -> bool {
+        if !self.look_ahead(1, |t| *t == token::CloseDelim(close_delim))
+            && self.eat(&token::DotDotDot)
+        {
+            // recover from typo of `...`, suggest `..`
+            let span = self.prev_token.span;
+            self.sess.emit_err(MissingDotDot { token_span: span, sugg_span: span });
+            return true;
+        }
+        false
     }
 
     /// Parses `ident (COLON expr)?`.

--- a/compiler/rustc_parse/src/parser/expr.rs
+++ b/compiler/rustc_parse/src/parser/expr.rs
@@ -2880,7 +2880,7 @@ impl<'a> Parser<'a> {
         };
 
         while self.token != token::CloseDelim(close_delim) {
-            if self.eat(&token::DotDot) || self.recover_struct_fileds_dots(close_delim) {
+            if self.eat(&token::DotDot) || self.recover_struct_field_dots(close_delim) {
                 let exp_span = self.prev_token.span;
                 // We permit `.. }` on the left-hand side of a destructuring assignment.
                 if self.check(&token::CloseDelim(close_delim)) {
@@ -3027,7 +3027,7 @@ impl<'a> Parser<'a> {
         self.recover_stmt();
     }
 
-    fn recover_struct_fileds_dots(&mut self, close_delim: Delimiter) -> bool {
+    fn recover_struct_field_dots(&mut self, close_delim: Delimiter) -> bool {
         if !self.look_ahead(1, |t| *t == token::CloseDelim(close_delim))
             && self.eat(&token::DotDotDot)
         {

--- a/src/test/ui/parser/issue-102806.rs
+++ b/src/test/ui/parser/issue-102806.rs
@@ -1,0 +1,22 @@
+#![allow(dead_code)]
+
+struct V3 {
+    x: f32,
+    y: f32,
+    z: f32,
+}
+
+fn pz(v: V3) {
+    let _ = V3 { z: 0.0, ...v};
+    //~^ ERROR expected `..`
+    //~| ERROR missing fields `x` and `y` in initializer of `V3`
+    let _ = V3 { z: 0.0, ... };
+    //~^ expected identifier
+    //~| ERROR missing fields `x` and `y` in initializer of `V3`
+
+    let _ = V3 { z: 0.0, ...Default::default() };
+    //~^ ERROR expected `..`
+    //~| ERROR missing fields `x` and `y` in initializer of `V3`
+}
+
+fn main() {}

--- a/src/test/ui/parser/issue-102806.rs
+++ b/src/test/ui/parser/issue-102806.rs
@@ -1,5 +1,6 @@
 #![allow(dead_code)]
 
+#[derive(Default)]
 struct V3 {
     x: f32,
     y: f32,
@@ -9,14 +10,16 @@ struct V3 {
 fn pz(v: V3) {
     let _ = V3 { z: 0.0, ...v};
     //~^ ERROR expected `..`
-    //~| ERROR missing fields `x` and `y` in initializer of `V3`
+
+    let _ = V3 { z: 0.0, ...Default::default() };
+    //~^ ERROR expected `..`
+
     let _ = V3 { z: 0.0, ... };
     //~^ expected identifier
     //~| ERROR missing fields `x` and `y` in initializer of `V3`
 
-    let _ = V3 { z: 0.0, ...Default::default() };
-    //~^ ERROR expected `..`
-    //~| ERROR missing fields `x` and `y` in initializer of `V3`
+    let V3 { z: val, ... } = v;
+    //~^ ERROR expected field pattern
 }
 
 fn main() {}

--- a/src/test/ui/parser/issue-102806.stderr
+++ b/src/test/ui/parser/issue-102806.stderr
@@ -1,0 +1,51 @@
+error: expected `..`, found `...`
+  --> $DIR/issue-102806.rs:10:26
+   |
+LL |     let _ = V3 { z: 0.0, ...v};
+   |                          ^^^
+   |
+help: use `..` to fill in the rest of the fields
+   |
+LL |     let _ = V3 { z: 0.0, ..v};
+   |                          ~~
+
+error: expected identifier, found `...`
+  --> $DIR/issue-102806.rs:13:26
+   |
+LL |     let _ = V3 { z: 0.0, ... };
+   |             --           ^^^ expected identifier
+   |             |
+   |             while parsing this struct
+
+error: expected `..`, found `...`
+  --> $DIR/issue-102806.rs:17:26
+   |
+LL |     let _ = V3 { z: 0.0, ...Default::default() };
+   |                          ^^^
+   |
+help: use `..` to fill in the rest of the fields
+   |
+LL |     let _ = V3 { z: 0.0, ..Default::default() };
+   |                          ~~
+
+error[E0063]: missing fields `x` and `y` in initializer of `V3`
+  --> $DIR/issue-102806.rs:10:13
+   |
+LL |     let _ = V3 { z: 0.0, ...v};
+   |             ^^ missing `x` and `y`
+
+error[E0063]: missing fields `x` and `y` in initializer of `V3`
+  --> $DIR/issue-102806.rs:13:13
+   |
+LL |     let _ = V3 { z: 0.0, ... };
+   |             ^^ missing `x` and `y`
+
+error[E0063]: missing fields `x` and `y` in initializer of `V3`
+  --> $DIR/issue-102806.rs:17:13
+   |
+LL |     let _ = V3 { z: 0.0, ...Default::default() };
+   |             ^^ missing `x` and `y`
+
+error: aborting due to 6 previous errors
+
+For more information about this error, try `rustc --explain E0063`.

--- a/src/test/ui/parser/issue-102806.stderr
+++ b/src/test/ui/parser/issue-102806.stderr
@@ -1,5 +1,5 @@
 error: expected `..`, found `...`
-  --> $DIR/issue-102806.rs:10:26
+  --> $DIR/issue-102806.rs:11:26
    |
 LL |     let _ = V3 { z: 0.0, ...v};
    |                          ^^^
@@ -9,16 +9,8 @@ help: use `..` to fill in the rest of the fields
 LL |     let _ = V3 { z: 0.0, ..v};
    |                          ~~
 
-error: expected identifier, found `...`
-  --> $DIR/issue-102806.rs:13:26
-   |
-LL |     let _ = V3 { z: 0.0, ... };
-   |             --           ^^^ expected identifier
-   |             |
-   |             while parsing this struct
-
 error: expected `..`, found `...`
-  --> $DIR/issue-102806.rs:17:26
+  --> $DIR/issue-102806.rs:14:26
    |
 LL |     let _ = V3 { z: 0.0, ...Default::default() };
    |                          ^^^
@@ -28,24 +20,26 @@ help: use `..` to fill in the rest of the fields
 LL |     let _ = V3 { z: 0.0, ..Default::default() };
    |                          ~~
 
-error[E0063]: missing fields `x` and `y` in initializer of `V3`
-  --> $DIR/issue-102806.rs:10:13
-   |
-LL |     let _ = V3 { z: 0.0, ...v};
-   |             ^^ missing `x` and `y`
-
-error[E0063]: missing fields `x` and `y` in initializer of `V3`
-  --> $DIR/issue-102806.rs:13:13
+error: expected identifier, found `...`
+  --> $DIR/issue-102806.rs:17:26
    |
 LL |     let _ = V3 { z: 0.0, ... };
-   |             ^^ missing `x` and `y`
+   |             --           ^^^ expected identifier
+   |             |
+   |             while parsing this struct
+
+error: expected field pattern, found `...`
+  --> $DIR/issue-102806.rs:21:22
+   |
+LL |     let V3 { z: val, ... } = v;
+   |                      ^^^ help: to omit remaining fields, use one fewer `.`: `..`
 
 error[E0063]: missing fields `x` and `y` in initializer of `V3`
   --> $DIR/issue-102806.rs:17:13
    |
-LL |     let _ = V3 { z: 0.0, ...Default::default() };
+LL |     let _ = V3 { z: 0.0, ... };
    |             ^^ missing `x` and `y`
 
-error: aborting due to 6 previous errors
+error: aborting due to 5 previous errors
 
 For more information about this error, try `rustc --explain E0063`.

